### PR TITLE
How to deploy easegress as a cluster

### PR DESCRIPTION
**Pull request by samutamm**
Add cookbook chapter about easegress cluster. Depends on https://github.com/megaease/easegress/pull/382

### CODE REVIEW

File: doc/cookbook/multi_node_cluster.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  - [Multiple instances in single node](#multiple-instances-in-single-node)
+  - [Multiple nodes](#multiple-nodes)
+  - [Readers and writers](#readers-and-writers)
+  - [References](#references)
+
+It is easy to start multiple Easegress instances to form an Easegress cluster, using `easegress-server` binary.
+
+##  Multiple instances in single node
+
+To start a single writer instance and two readers, we need to run start the server in three separate terminals. It's enough to define `localhost` for `cluster-join-urls` value and leave most of the options to default values.
```
Comment (localvar) [inline doc/cookbook/multi_node_cluster.md:14]:
```suggestion
To start a single writer instance and two readers, we need to start the server in three separate terminals. It's enough to define `localhost` for `cluster-join-urls` value and leave most of the options to default values.
```

File: doc/cookbook/multi_node_cluster.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  --cluster-listen-peer-urls http://$HOST4:2374 \
+  --cluster-listen-client-urls http://$HOST4:2373 \
+  --cluster-advertise-client-urls http://$HOST4:2373 \
+  --cluster-join-urls http://$HOST1:2380,http://$HOST2:2378,http://$HOST3:2376,http://$HOST4:2374
+```
+
+
+##  Readers and writers
+
+When running Easegress as a cluster, each instance has either *writer* or *reader* role. There should be odd number off *writers* (1,3,5,7,9) for [failure tolerance](https://etcd.io/docs/v3.5/faq/#what-is-failure-tolerance) and any number of *readers*.
```
Comment (localvar) [inline doc/cookbook/multi_node_cluster.md:131]: we need to clarify the relationship between Easegress roles and Etcd roles, or people may get confused why we reference an Etcd document here.

Comment (localvar) [reply doc/cookbook/multi_node_cluster.md:131]: And all the examples above have only one `writer`, propose to add an example for multiple writers.

**Review by localvar — APPROVED** [review]

**Comment by suchen-sci** [comment]
Please remember to update `README.md` `##User Cases`, `README.zh-CN.md` `## 用户案例`  and `doc/cookbook/README.md`

File: doc/cookbook/multi_node_cluster.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+memory-profile-file:
+log-dir: ./log
+debug: false
+cluster:
+  primary-listen-peer-urls: http://$HOST1:2380
+```
+
+##  Cluster role
+
+When running Easegress as a cluster, each instance has either *primary* or *secondary* role. *Primary* nodes persist the Easegress state on the disk, while *secondary* request this information from their peers (defined by `initial-advertise-peer-urls` parameter).
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:138]: Please add a note to describe what is peers here. That will be better.

File: doc/cookbook/multi_node_cluster.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+It is a good practice to choose an odd number (1,3,5,7,9) of *primary* nodes, to tolerate failures of *primary* nodes. This way the cluster can stay in healthy state, even if the network partitions. With an even number of *primary* nodes, the cluster can be divided to two groups of equal size due to network partition. Then neither of the sub-cluster have the majority required for consensus (of Raft algorithm). However with odd number of *primary* nodes, the cluster cannot be divided to two groups of equal size and this problem cannot occur.
+
+For the *secondary* nodes, there is no constraints for the number of nodes. Secondary nodes do not participate consensus vote of the cluster, so their failure do not affect the cluster health. Adding more (*secondary*) nodes does still increase the communication between nodes.
+
+### Cluster roles in etcd terminology
```
[the remark this answers is not in the document]
Comment (samutamm) [reply doc/cookbook/multi_node_cluster.md]: I separated the practical tutorial part and Q&A type of detail / theory section. There are now two diagrams to visualize the concepts.

File: doc/cookbook/multi_node_cluster.md [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+It is easy to start multiple Easegress instances to form an Easegress cluster. This tutorial provides instructions how to create stand-alone Easegress cluster by starting multiple Easegress instances.
+
+## Prerequisite
+
+The following prerequisites are required for a successful deployment of Easegress cluster.
+- latest `easegress-server` and `egctl` binaries (run `make` in root of the repository)
+- successful creation of an Easegress pipeline (like the Hello World example in the README.md of the repository or any other chapter in doc/cookbook)
+- few machines that are in the same network or otherwise accessible or Docker or other container technology. If you only have one machine, then you could use localhost as the host and modify the ports in the example.
+
+## Example
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:29]:
```suggestion
## Deploy an Easegress cluster step by step
```

File: doc/cookbook/multi_node_cluster.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+wal-dir: ""
+cpu-profile-file:
+memory-profile-file:
+log-dir: ./log
+debug: false
+cluster:
+  primary-listen-peer-urls: http://$HOST1:2380
+```
+
+## Configuration tips (optional)
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:181]:
```suggestion
## Configuration tips
```

File: doc/cookbook/multi_node_cluster.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+Here's a drawing that illustrates the difference.
+
+<p align="center">
+  <img src="./easegress-cluster-connections.png" width=300>
+</p>
+
+*Primary* members need to synchronize with peers (other *primary* members). Meanwhile *secondary* can read and update the state through one *primary* member.
+
+*How are etcd server and primary related?*
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:216]:
```suggestion
*How are etcd server and Easegress primary related?*
```

File: doc/cookbook/multi_node_cluster.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+It is a good practice to choose an odd number (1,3,5,7,9) of *primary* nodes, to tolerate failures of *primary* nodes. This way the cluster can stay in healthy state, even if the network partitions. With an even number of *primary* nodes, the cluster can be divided to two groups of equal size due to network partition. Then neither of the sub-clusters have the majority required for consensus. However with odd number of *primary* nodes, the cluster cannot be divided to two groups of equal size and this problem cannot occur.
+
+For the *secondary* nodes, there is no constraints for the number of nodes. Secondary nodes do not participate consensus vote of the cluster, so their failure do not affect the cluster health. Adding more (*secondary*) nodes does still increase the communication between nodes.
+
+ *Can number of primary members scale up?*
+
+Please note that it is not recommended to add additional node with `primary` cluster role, but `primary` nodes should be started at cluster start up. When scaling up the cluster, it is recommended to add and remove `secondary` cluster members.
+
+*What are advertise-peer-urls and listen-client-urls?*
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:193]:
```suggestion
*What are `advertise-peer-urls` and `listen-client-urls`?*
```

File: doc/cookbook/multi_node_cluster.md [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+<p align="center">
+  <img src="./easegress-cluster-connections.png" width=300>
+</p>
+
+*Primary* members need to synchronize with peers (other *primary* members). Meanwhile *secondary* can read and update the state through one *primary* member.
+
+*How are etcd server and primary related?*
+
+Easegress cluster uses [etcd](https://etcd.io) distributed key-value store to synchronize the cluster state. The primary and secondary cluster roles have following relation with `etcd`:
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:218]:
```suggestion
Easegress uses [etcd](https://etcd.io) distributed key-value store to synchronize the cluster state. The primary and secondary cluster roles have the following relation with `etcd`:
```

File: doc/cookbook/multi_node_cluster.md [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+| Easegress cluster role   | primary   | secondary   |
+|-----|-----|-----|
+| etcd term | server | client |
+
```
Comment (jacksonwuu) [inline doc/cookbook/multi_node_cluster.md:223]:
```suggestion
For a better understanding, you can think `easegress primary` is an `etcd server`, and `easegress secondary` is a `etcd client`.
```

**Review by jacksonwuu — COMMENTED** [review]
Great work! I have added some small and not-so-matters suggestions, and you can choose which one to accept, and make a request again, then I will approve this PR.

File: doc/cookbook/multi_node_cluster.md (hunk 6 above)
Comment (samutamm) [reply doc/cookbook/multi_node_cluster.md:181]: Thanks for the comments! I'll keep `optional` here as this is not required for the tutorial.

**Review by jacksonwuu — APPROVED** [review]

**Review by suchen-sci — APPROVED** [review]

**Pull request merged**